### PR TITLE
build(aio): render subclasses in API docs

### LINKED ISSUE #19306: Documentation doesn't show all the inherited public and private methods and attributes
## I'm submitting a...
[x] Documentation issue or request

## Current behavior
For classes that extend other classes, there is no way see all the public and protected methods and attributes.

## Expected behavior
Please include a mechanism to show/hide all the inherited public and protected methods and attributes of a class. This particularly important for deep inheritance trees.

## What is the motivation / use case for changing the behavior?
Other languages that support inheritance have their documentation include a mechanism to show/hide all the inherited public and protected methods and attributes.

## Environment
Angular version: 4.4.3

Browser:
- [x] Chrome (desktop) version 61.0.3163.91

### PR DESCRIPTION
**Pull request by petebacondarwin**
Closes #19306

### CODE REVIEW

**Comment by petebacondarwin** [comment]
See https://pr19343-567b1c7.ngbuilds.io/api/forms/AbstractControlDirective for an example

File: aio/tools/transforms/templates/api/lib/descendants.html [hunk 1]
```diff
@@ -0,0 +1,12 @@
+{% macro renderDescendants(doc) %}
+  {% if doc.descendants.length %}
+  <ul>
+  {% for descendant in doc.descendants %}
+    <li>
+      <pre class="prettyprint lang-ts"><code>{$descendant.name $}</code></pre>
```
Comment (gkalpak) [inline aio/tools/transforms/templates/api/lib/descendants.html]: `{$desc...` --> `{$ desc...` :smiley:

**Review by gkalpak — APPROVED** [review]
Looks great. I might consider making the `ul` more compact and making the whole section expandable/collapseable (and maybe collapsed by default), but not necessarily on this PR.

LGTM 👍

**Comment by petebacondarwin** [comment]
Still a WIP - I have some other ideas that need to be included here - and some e2e tests to ensure it works (and continues to work).

**Comment by petebacondarwin** [comment]
OK, so ready for a review now.

File: aio/tools/transforms/angular-base-package/rendering/filterByPropertyValue.spec.js [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+describe('filterByPropertyValue filter', () => {
+  let filter;
+
+  beforeEach(function() { filter = factory(); });
+
+  it('should be called "filterByPropertyValue"', function() { expect(filter.name).toEqual('filterByPropertyValue'); });
+
+  it('should filter out items that do not match the given property value', function() {
+    expect(filter.process([{ a: 1 }, { a: 2 }, { b: 1 }, { a: 1, b: 2}, { a: null }, { a: undefined }], 'a', 1))
```
Comment (gkalpak) [inline aio/tools/transforms/angular-base-package/rendering/filterByPropertyValue.spec.js]: Missing space in `b: 2}` 😞

File: aio/tools/transforms/templates/api/interface.template.html [hunk 3]
```diff
@@ -1,9 +1,12 @@
 {% import "lib/paramList.html" as params -%}
 {% import "lib/memberHelpers.html" as memberHelper -%}
+{% import "lib/descendants.html" as descendants -%}
 {% extends 'export-base.template.html' -%}
 
 {% block overview %}{% include "includes/interface-overview.html" %}{% endblock %}
 {% block details %}
   {% include "includes/description.html" %}
+  {$ descendants.renderDescendants(doc, 'interface', 'Child Interfaces') $}
+  {$ descendants.renderDescendants(doc, 'class', 'Class Implementations') $}
```
Comment (gkalpak) [inline aio/tools/transforms/templates/api/interface.template.html]: ❤️

File: aio/tools/transforms/templates/api/lib/descendants.html [hunk 4]
```diff
@@ -0,0 +1,14 @@
+{% macro renderDescendants(doc, docType, title='', recursed=false) %}
+  {% set descendants = doc.descendants | filterByPropertyValue("docType", docType) %}
```
Comment (gkalpak) [inline aio/tools/transforms/templates/api/lib/descendants.html]: `"docType"` --> `'docType'` (?) :grin:

File: aio/e2e/api.po.ts [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    // ul
+    //  li
+    //    code
+    //    ul
+    //      li
+    //        code
+    //    ul
+    //      li
+    //        code
+    //  li
```
Comment (gkalpak) [inline aio/e2e/api.po.ts]: These 1st level `li` are not indented correctly 😞

File: aio/e2e/api.po.ts [hunk 6]
```diff
@@ -0,0 +1,25 @@
+import { browser, element, by, promise, ElementFinder, ExpectedConditions } from 'protractor';
```
Comment (gkalpak) [inline aio/e2e/api.po.ts]: Several unused imports.

File: aio/e2e/api.e2e-spec.ts [hunk 7]
```diff
@@ -0,0 +1,29 @@
+import { browser, element, by, promise } from 'protractor';
+import { ApiPage } from './api.po';
+
+describe('Api pages', function() {
+  it('should show direct subclasses of a class', () => {
+    const page = new ApiPage();
```
Comment (gkalpak) [inline aio/e2e/api.e2e-spec.ts]: Nit: This is a good candidate for `beforeEach`'ing.

File: aio/e2e/api.po.ts [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+import { SitePage } from './app.po';
+
+export class ApiPage extends SitePage {
+  getSubClasses(docType: string, onlyDirect = false) {
```
Comment (gkalpak) [inline aio/e2e/api.po.ts]: Nit: Maybe call the method `getDescendants`.

**Review by gkalpak — APPROVED** [review]
LGTM 👍 
(Just a few of nits (because...why not :grin:))

**Comment by gkalpak** [comment]
(I assume this is safe for patch too, right?)

File: aio/tools/transforms/templates/api/lib/descendants.html (hunk 4 above)
Comment (petebacondarwin) [reply aio/tools/transforms/templates/api/lib/descendants.html]: I was sure that nunjucks didn't accept single quotes but I appear to be wrong... 

https://mozilla.github.io/nunjucks/templating.html#expressions

Perhaps it was just my IDE plugin that was incorrectly syntax highlighting it.

Well you learn something new everyday.

File: aio/e2e/api.po.ts (hunk 6 above)
Comment (petebacondarwin) [reply aio/e2e/api.po.ts]: And even more in the spec file!

File: aio/e2e/api.e2e-spec.ts (hunk 7 above)
Comment (petebacondarwin) [reply aio/e2e/api.e2e-spec.ts]: But I really don't want to use `beforeEach` here 😠 
So I am going to fly in the face of convention and pass the URL to the constructor!!

Comment (petebacondarwin) [reply aio/e2e/api.e2e-spec.ts]: (because why would you want an API page object that has not navigated to the page?)

**Comment by petebacondarwin** [comment]
Good for master and patch yes

**Pull request merged**